SWIFT-731 Update guides to focus on async API

JamesHeppenstall opened the pull request.

kmahar commented:
I agree we should aim to make the examples appear as "real world" as possible, and in apps people generally won't be using `wait()`.

JamesHeppenstall commented:
I've modified the change stream examples and reverted the error handling ones back to their original form. Let me know if you want me to change the error handling examples to async versions or keep them as sync.

patrickfreed commented on Guides/Change-Streams.md:
```diff
@@ -8,158 +8,155 @@ MongoSwift 0.2.0 added support for [change streams](https://docs.mongodb.com/man
-    // process `ChangeStreamEvent<Document>` here
+inventory.watch().flatMap { stream in // a `ChangeStream<ChangeStreamEvent<Document>>`
+	// read the first change event
+	stream.next().whenComplete{ _ in
```
instead of using `whenComplete` here, we can just use chains of `flatMap`. That way, the top level `whenFailure` will handle any errors that occur anywhere in the callback chain.

JamesHeppenstall replied:
Done.

JamesHeppenstall commented on Guides/Change-Streams.md:
```diff
@@ -8,158 +8,155 @@ MongoSwift 0.2.0 added support for [change streams](https://docs.mongodb.com/man
-for change in stream {
-    // process `ChangeStreamEvent<Document>` here
+inventory.watch().flatMap { stream in // a `ChangeStream<ChangeStreamEvent<Document>>`
+	stream.forEach { event in
```
Ok I've changed all indentations to spaces.

JamesHeppenstall commented on Guides/Change-Streams.md:
```diff
@@ -8,158 +8,155 @@ MongoSwift 0.2.0 added support for [change streams](https://docs.mongodb.com/man
-let resumedStream = try inventory.watch(options: ChangeStreamOptions(resumeAfter: resumeToken))
-for change in resumedStream {
-    // process `ChangeStreamEvent<Document>` here
+inventory.watch().flatMap { stream in // a `ChangeStream<ChangeStreamEvent<Document>>`
```
I've done this but not sure if it's the way you were thinking. Let me know what you think!

patrickfreed replied:
It's really close, but it's a little racy because it doesn't guarantee the resume attempt happens after killing the stream. The `stream.next()` invocation returns immediately and starts some operations in the background, and the next line we check for the session id. If `next` hasn't completed by the time we check for the id, then it will still be `nil`.

To alleviate this, we can just use `flatMap` chains to guarantee they will be ordered (this in one line for demonstration purposes):
```swift
stream.next().flatMap { _ in stream.kill() }.flatMap { /* resume */ }
```
Using a single `flatMap` chain also has the added benefit that the single `whenFailure` will catch all errors. In the existing implementation, we don't have any callbacks handling the errors of `next` or `kill` since we're only returning the future from `inventory.watch`.

JamesHeppenstall replied:
Got it - I've made the necessary changes!

kmahar commented:
LGTM as well, ditto patrick's comment about updating the title/commit message. 

though I do wonder if it is worth keeping around the sync version of the guide as well in a separate file since we do have a synchronous API still, thoughts?
